### ISSUE TITLE
Flow doesn't like defaulting arguments to `this.foo`

### ISSUE THREAD

**Issue by zyklus** [facebook/flow#3353]
I'm not sure what the spec says about this, but flow and babel are out of sync here.

```js
class Foo {
  a = 42;
  bar(a = this.a) {
    console.log(a);
  }
}

new Foo().bar()
```

Flow claims that in the assignment to `this.a` that `this` is global.  Babel transpiles it to the class instance.

**Comment by AugustinLF** [comment]
It should be fixed quite soon, there's an open PR #3334 to fix that!

**Comment by jeffmo** [comment]
For posterity, here's my breakdown of the spec confirming Babel's behavior:

* [12.3.4.1](https://tc39.github.io/ecma262/#sec-function-calls-runtime-semantics-evaluation), Step 7.a.i gets the `thisValue` of a `.`-ed member expression
* Step 11 returns [`EvaluateDirectCall(func, thisValue, arguments, tailCall).`](https://tc39.github.io/ecma262/#sec-evaluatedirectcall)
* Step 5 of `EvaluateDirectCall` delegates to [`Call ( F, V [ , argumentsList ] )`](https://tc39.github.io/ecma262/#sec-call)
* `Call` runs the internal [`[[Call]]`](https://tc39.github.io/ecma262/#sec-ecmascript-function-objects-call-thisargument-argumentslist) method of the function
* Step 4 of `[[Call]]` runs [`PrepareForOrdinaryCall`](https://tc39.github.io/ecma262/#sec-prepareforordinarycall) (which is what sets the `[[HomeObject]]` for use by `super` by way of [`NewFunctionEnvironment`](https://tc39.github.io/ecma262/#sec-newfunctionenvironment))
* Step 6 of `[[Call]]` -> [`OrdinaryCallBindThis ( F, calleeContext, thisArgument )`](https://tc39.github.io/ecma262/#sec-ordinarycallbindthis) which sets `this` in the lexical environment
* Step 7 of `[[Call]]` -> [`OrdinaryCallEvaluateBody ( F, argumentsList )`](https://tc39.github.io/ecma262/#sec-ordinarycallevaluatebody)
* `OrdinaryCallEvaluateBody` calls [`EvaluateBody`](https://tc39.github.io/ecma262/#sec-function-definitions-runtime-semantics-evaluatebody)
* `EvaluateBody` calls [`FunctionDeclarationInstantiation`](https://tc39.github.io/ecma262/#sec-functiondeclarationinstantiation)
* [9.2.12 - FunctionDeclarationInstantiation ( func, argumentsList )](https://tc39.github.io/ecma262/#sec-functiondeclarationinstantiation), Step 27 is where we evaluate defaults as part of a function call.
* Step 27.b creates a new lexical environment with a parent equal to the outer environment
* Step 27.f then goes on to evaluate the defaults in this new environment
